### DISCUSSION TITLE
Upgrading from 417 > 418+ breaks proxy host for the NPM+ web panel

### DISCUSSION THREAD

**Discussion by DessertOverlord** [ZoeyVid/NPMplus#1278] [Q&A] [answered]
For a few months I hosted NPM on an RPi5, and recently I have been migrating all of my containers over to my x86 server PC. When I did so, I accidentally updated NPM+ to latest, up from v417 on Dockerhub. The only thing that appears to have broken is that previously I had a proxy host that went from https://npm.mydomain.tld/ to https://127.0.0.1:81, so that I could remotely access the webUI. 

I use Cloudflare's tunnels and SSL, and my setup was working in v417, but updating to v418 irreversibly breaks it. Navigating to https://npm.mydomain.tld/ enters an infinite redirect loop. I can also confirm that the "Force HTTPS" option is disabled. 

Even if I tear the container down and rebuild on v417, npmplus throws an error in the logs, which implies that the breaking change lives somewhere in my persistent data:

`
npmplus  | nginx: [emerg] open() "/usr/local/nginx/conf/conf.d/include/tls-ciphers-no-stapling.conf" failed (2: No such file or directory) in /data/nginx/default.conf:19
npmplus  | nginx: configuration file /usr/local/nginx/conf/nginx.conf test failed
`

The only way I have gotten this to work again is by reverting to my original backup on v417.

**Comment by Zoey2936** [comment]
yes downgrading is not supported. Can you please share the config file of the host redirecting in loop?

**Comment by DessertOverlord** [comment]
Here is the config, I only scrubbed my own domain names from it.

```
# ------------------------------------------------------------
# npm.mydomain.tld
# ------------------------------------------------------------


server {
    set $forward_scheme https;
    set $server "127.0.0.1";
    set $port 81;

    listen unix:/run/nginx-2.sock;

    listen 0.0.0.0:80;
    listen [::]:80;


    listen 0.0.0.0:443 ssl;
    listen [::]:443 ssl;

    listen 0.0.0.0:443 quic;
    listen [::]:443 quic;


    more_set_headers 'Alt-Svc: h3=":443"; ma=86400';


    server_name npm.mydomain.tld;


    # Certbot TLS
    include conf.d/include/tls-ciphers.conf;
    ssl_certificate /data/tls/certbot/live/npm-2/fullchain.pem;
    ssl_certificate_key /data/tls/certbot/live/npm-2/privkey.pem;
    ssl_trusted_certificate /data/tls/certbot/live/npm-2/chain.pem;


    # Enable Brotli
    include conf.d/include/brotli.conf;


    include conf.d/include/always.conf;


    # custom locations
    location / {
        include conf.d/include/always.conf;


        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade;


        # Proxy!
        include conf.d/include/proxy.conf;
    }


    # Custom
    include /data/nginx/custom/server_proxy.conf;
}
```

**Reply by Zoey2936** [reply]
and this config leads to a redirect loop?

**Reply by DessertOverlord** [reply]
For some reason it does when used in conjunction with Cloudflare's tunnels, but only in versions 418+. In version 417, it works exactly as intended with the same setup.

**Reply by Zoey2936** [reply]
I think I know the reason, does crowdsec proxy to http or https?

**Reply by Zoey2936** [reply]
because of: https://github.com/ZoeyVid/NPMplus/pull/1025

**Reply by Zoey2936** [reply]
It should be fixable by changes your setting in the cloudflare dashboard

**Reply by DessertOverlord** [reply]
Hmm, what setting do I need to change? I have SSL set to Full (strict) and the forced HTTPS rewrites disabled right now. Is there something else I'm missing?

**Reply by Zoey2936** [reply]
can you please send screenshots of your cloudflare config?

**Reply by DessertOverlord** [reply]
![image](https://github.com/user-attachments/assets/65289797-fea8-4fd9-ab45-5e0a307f72dc)
![image](https://github.com/user-attachments/assets/bffea899-99e2-44e2-a552-5690f09398c0)
![image](https://github.com/user-attachments/assets/6c2e1175-9367-40ec-82e2-241f067493cf)

That's most of the SSL config I think. Let me know if you want to see anything else.

**Reply by Zoey2936** [reply]
can you revert the changes done in the mentioned PR, by editing the file inside the container (nano and bash are installed in the container) so `docker exec -it npmplus nano /usr/local/nginx/conf/conf.d/include/force-tls.conf` should work. Then please rsdtart the contaienr and report back

**Reply by DessertOverlord** [reply]
I commented out lines 8-10 in the file so the contents are now:
```
Check if the original scheme is HTTP
if ($scheme = "http") {
    return 301 https://$host$request_uri;
}

# Check if the request was forwarded with HTTP protocol
# This is necessary when behind a proxy like Cloudflare
#if ($http_x_forwarded_proto = "http") {
#    return 301 https://$host$request_uri;
#}
```

and that seems to work on version 2.12.1+9428805 as reported by the container. I also did a 
`docker compose pull
[+] Pulling 1/1
 ✔ npmplus Pulled`
to ensure I was on latest. 

The repo digest is sha256:6f6a28e0f4978bb06653ab3df335d1c31295b4f2c4313dcc02c6cdbe7df31644 just in case.

**Reply by Zoey2936** [reply]
so was it fixed by updating or changing the file?

**Reply by DessertOverlord** [reply]
Yes, seems like it. It's weird that it only affects my proxy host pointing towards the NPM panel itself.

**Reply by Zoey2936** [reply]
so changing the file or updating?

**Reply by DessertOverlord** [reply]
Oh sorry I misunderstood. I didn't test it with only updating this time, but previously that is what broke it. Changing the file on latest makes it work right now. I'll do a little bit more testing in a few minutes

**Reply by DessertOverlord** [reply]
When navigating to my proxy host after a clean run:
![image](https://github.com/user-attachments/assets/d37ca5c0-7816-415c-874d-7aac8212fd18)

I made these changes:
- `docker exec -it npmplus nano /usr/local/nginx/conf/conf.d/include/force-tls.conf`
- Comment out lines 8-10
- `docker compose stop`
- `docker compose start`

Now navigating to the same URL (I just CTRL+SHIFT+R'd):
![image](https://github.com/user-attachments/assets/8f87278e-d15e-40e9-9c65-9ac6b2381f43)

So whatever happened in #1025 bricked my NPM panel through CF tunnels, apparently.

**Reply by Zoey2936** [reply]
something in your chain triggers http_x_forwarded_proto, so something uses http which it should not

**Reply by Zoey2936** [reply]
hoiw did you configured your tunnels?

**Reply by DessertOverlord** [reply]
What's really odd about this is that I do _not_ have Force HTTPS enabled, so why is rootfs/usr/local/nginx/conf/conf.d/include/force-tls.conf being invoked?

![image](https://github.com/user-attachments/assets/da419216-2de2-4ab3-b04d-183a0eb6a4de)

![image](https://github.com/user-attachments/assets/233c83d7-dba2-4a9a-b61a-f60b74f4fa2f)

**Reply by DessertOverlord** [reply]
> hoiw did you configured your tunnels?

![image](https://github.com/user-attachments/assets/b9089680-cde7-42a6-9d26-d1eac1a8a5c7)

**Reply by Zoey2936** [reply]
it is included by NPMplus own config: https://github.com/ZoeyVid/NPMplus/blob/f8a7ff0f6367467256854ac89faa114a358c1305/rootfs/usr/local/nginx/conf/conf.d/npm.conf#L8

**Reply by Zoey2936** [reply]
> > hoiw did you configured your tunnels?
> 
> ![image](https://private-user-images.githubusercontent.com/16036856/392133166-b9089680-cde7-42a6-9d26-d1eac1a8a5c7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyNTkxNjQsIm5iZiI6MTczMzI1ODg2NCwicGF0aCI6Ii8xNjAzNjg1Ni8zOTIxMzMxNjYtYjkwODk2ODAtY2RlNy00MmE2LTlkMjYtZDFlYWMxYThhNWM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjAzVDIwNDc0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI4NGM2MGNjMTA5NDZkYzdhODYyMWVjYzZkZjVmZjE3OTU4Y2YxMGZkMDEyZTYzMTQ0MmNlNzIxZTJhMjE4NzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.F6buqxwJlMSBycgMBB99qGhLUlrizGw1V0qqreF99pw)

there is your issue, please try to use https

**Reply by DessertOverlord** [reply]
I did a `docker compose down` and `docker compose up -d` to refresh the container, and validated I still got the loop. 

Then I changed to this new CF Tunnel config:
![image](https://github.com/user-attachments/assets/57034598-b178-4723-8c27-d0f0997d8186)

Result:
![image](https://github.com/user-attachments/assets/462d538b-b326-4f5f-bc15-0a669fb7b705)

**Reply by Zoey2936** [reply, marked as answer]
try this (additionally): https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/configure-tunnels/origin-configuration/#notlsverify

**Reply by DessertOverlord** [reply]
Mint, that did it. All is well.

So in summary, that means the issue was that Cloudflare was sending HTTP traffic to NPM, which it was trying to redirect to HTTPS because the NPM+ panel has the force-tls included by default? Either way, thank you very much for the help!

**Reply by Zoey2936** [reply]
yes exactly
